awx/ui/src/screens/TopologyView/MeshGraph.js

```diff
@@ -0,0 +1,292 @@
+import * as d3 from 'd3';
+import Legend from './Legend';
+import Tooltip from './Tooltip';
+import ContentLoading from './ContentLoading';
+import {
+  renderStateColor,
+  renderLabelText,
+  renderNodeType,
+  renderNodeIcon,
+  redirectToDetailsPage,
+  // generateRandomNodes,
+  // getRandomInt,
+} from './utils/helpers';
+import {
+  MESH_FORCE_LAYOUT,
+  DEFAULT_RADIUS,
+  DEFAULT_NODE_COLOR,
+  DEFAULT_NODE_HIGHLIGHT_COLOR,
+  DEFAULT_NODE_LABEL_TEXT_COLOR,
+  DEFAULT_FONT_SIZE,
+  MARGIN,
+  HEIGHT,
+  FALLBACK_WIDTH,
+} from './constants';
+
+const Loader = styled(ContentLoading)`
+  height: 100%;
+  position: absolute;
+  width: 100%;
+  background: white;
+`;
+function MeshGraph({ data, showLegend, zoom }) {
+  // function MeshGraph({ showLegend, zoom }) {
+  const [isNodeSelected, setIsNodeSelected] = useState(false);
+  const [selectedNode, setSelectedNode] = useState(null);
+  const [nodeDetail, setNodeDetail] = useState(null);
+  const history = useHistory();
+
+  // const data = generateRandomNodes(getRandomInt(4, 50));
+  const draw = () => {
+    const getWidth = () => {
+      let width;
+      // This is in an a try/catch due to an error from jest.
+      // Even though the d3.select returns a valid selector with
+      // style function, it says it is null in the test
+      try {
+        width =
+          parseInt(d3.select(`#chart`).style('width'), 10) - MARGIN ||
+          FALLBACK_WIDTH;
+      } catch (error) {
+        width = FALLBACK_WIDTH;
+      }
+
+      return width;
+    };
+    const width = getWidth();
+
+    /* Add SVG */
+    d3.selectAll(`#chart > svg`).remove();
+    const svg = d3
+      .select('#chart')
+      .append('svg')
+      .attr('class', 'mesh-svg')
+      .attr('width', `${width + MARGIN}px`)
+      .attr('height', `${HEIGHT + MARGIN}px`)
+      .attr('viewBox', [0, 0, width, HEIGHT]);
+    const mesh = svg
+      .append('g')
+      .attr('class', 'mesh')
+      .attr('transform', `translate(${MARGIN}, ${MARGIN})`);
+
+    const graph = data;
+
+    const simulation = d3
+      .forceSimulation(graph.nodes)
+      .force(
+        'charge',
+        d3
+          .forceManyBody(MESH_FORCE_LAYOUT.defaultForceBody)
+          .strength(MESH_FORCE_LAYOUT.defaultForceStrength)
+      )
+      .force(
+        'link',
+        d3.forceLink(graph.links).id((d) => d.hostname)
+      )
+      .force(
+        'collide',
+        d3.forceCollide(MESH_FORCE_LAYOUT.defaultCollisionFactor)
+      )
+      .force('forceX', d3.forceX(MESH_FORCE_LAYOUT.defaultForceX))
+      .force('forceY', d3.forceY(MESH_FORCE_LAYOUT.defaultForceY))
+      .force('center', d3.forceCenter(width / 2, HEIGHT / 2));
+
+    const link = mesh
+      .append('g')
+      .attr('class', `links`)
+      .attr('data-cy', 'links')
+      .selectAll('line')
+      .data(graph.links)
+      .enter()
+      .append('line')
+      .attr('class', (_, i) => `link-${i}`)
+      .attr('data-cy', (d) => `${d.source}-${d.target}`)
```

Comment:
![image](https://user-images.githubusercontent.com/12464858/154345599-79ba4dce-b323-4b9e-9841-8e85659cb4bd.png)
```suggestion
      .attr('data-cy', (d) => `${d.source.hostname}-${d.target.hostname}`)
```